### DISCUSSION TITLE
Getting RX packets' header

### DISCUSSION THREAD

**Discussion by daitarn4** [corundum/corundum#217] [Q&A]
Hi,

I'm a bit new on getting under the hood of a Linux network device driver. For a project we deployed the corundum NIC on a ZCU102. Now, I gone in deep in understanding the driver code (first reading the documentation and then looking directly the code); I understood that the _mqnic_process_rx_cq()_ function is responsible to make the received packet data available to the network stack through the a _skb_ buffer. Now, I understood that the skb has a data pointer to the received packet (ethernet frame); however, I tried to print the source/dest MAC addresses, but it seems I just accessing to random (or not initialized) data. 

I tried this by passing the skb to the _eth_hdr()_ function and printing h_source/h_dest bytes at different points in the _mqnic_process_rx_cq()_ function (**mqnic_rx.c**)without being successful (i.e.,in the mqnic_rx.c file after line 346, after line 363, after line 379).

I was wondering if there is a better point (different function) where to get such information, or where I was getting wrong.

**Comment by alexforencich** [comment]
SKBs are a bit complicated.  Corundum receive works with bare pages, attaching the pages to SKBs right before hand-off to the stack.  So the data, at least initially, is not "in" the SKB, it's in the first attached page.  At some point, linux might pull some header components into the SKB.  I'm not sure offhand how to properly access the data inside the SKB, but there should be some linux APIs for that with take care of the details.

**Reply by daitarn4** [reply]
Dear Alex,

many thanks for your reply; indeed, we progressed in that direction. We basically followed your suggestion and coupled the Corundum driver with another kernel module (that lives on top of the driver) that performs some packet header processing. Beside this, we also extended a bit the Corundum driver functionalities including the possibility to setup the MAC addresses of the interface at runtime (it seems that the Corundum driver uses random ones that changes every time one needs to reboot the boards). 
We would also wonder if there are some example of how to implement an "app" that should live within Corundum, since we would like to perform the processing we are currently running on the extra-kernel module right on the hardware (e.g., inspect the packet headers).

Many thanks in advance.
